### PR TITLE
fix: Modified on page load actions calculation using AST

### LINKED ISSUE #9864: [Bug]: OnPageLoad Query not working when widget/Query name and data binding key name are same
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

For a user application shared on discord 

http://95.217.192.133:82/applications/6193655e2c16ce420ad195cd/pages/6198d7792c16ce420ad19656/edit?id=3

This particular page should logically have `layoutOnLoadActions` but when checked in the page response found layoutOnLoadActions as an empty array.

### Steps To Reproduce

Quoting @mohanarpit 

"There is a known edge case where if any widget is bound to a data field with the same name, the server thinks there is a cyclic dependency. This is because the server doesn’t yet do an AST (which the client has started to do).
Eg: Widget Name: monitor_name
Binding: {{ Api1.data.monitor_name }}
This will cause issues. Not the easiest thing to fix right now without introducing AST parsing of JS bindings on the server."

### Environment

Production

### Version

Self-Hosted

### LINKED ISSUE #16338: [Task]: Consume AST apis on backend server
### Is there an existing issue for this?

- [X] I have searched the existing issues

### SubTasks

The apis are added in RTS server. There are two apis for the AST backend usage :

- **Single script Identifiers**
  _Method_: `POST`
  _URL_: `/rts-api/v1/ast/single-script-data`
  _Body_: `{ "script": "(function(){ let a = 1; return [Api1.data, Api2.data, a]; })()", "evalVersion": 2 }` 
  _Response_: `{"success":true,"data":{"identifiers":["Api1.data","Api2.data"]}}`
  
- **Multiple script Identifiers**
  _Method_: `POST`
  _URL_: `/rts-api/v1/ast/multiple-script-data`
  _Body_: `{ "scripts": ["(function(){ let a = 1; return [Api1.data, Api2.data, a]; })()", "(function(){ let a = 1; return [XYZ.data, a]; })()"], "evalVersion": 2 }` 
  _Response_: `{"success":true,"data":[{"identifiers":["Api1.data","Api2.data"]},{"identifiers":["XYZ.data"]}]}`

> Where evalVersion is optional and default value is 2,  eval  version is the version we can find in the response of `/api/v1/pages?pageId=${pageId}` request of the application  where the path for it is `data.application. evaluationVersion`.

### LINKED ISSUE #17190: [Bug]: JSObj Sync() | Multiple toasts notifying its OnPageLoad run
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

https://user-images.githubusercontent.com/91450662/193046404-d460a0dd-25a8-4ab3-a4f6-03ce5d0f2952.mov

### Steps To Reproduce

1. Create a JSObject sync method to return a query data, here : '`return get_wikis.data`'
2. Bind this js sync function to some widget, here: table to display the response - '`{{JSObject1.myFun1.data}}`'
3. Create other independent queries & map them to any other widgets
4. **Observe: Along with the new notification toast, the JsObj.myFunc1 tags along & occurs multiple times**
5. **Expected: While JSObj.myFunc1 sync method is not expected in such cases**

### Public Sample App

_No response_

### Version

Prod/Release v1.8.4

### PR DESCRIPTION
**Pull request by nidhi-nair**
## Description
This PR introduces a new way of parsing all the dynamic bindings present in an Appsmith application. 

Earlier, we used to break down the string representation of a binding into words and try to predict what possible entities we were referring to in any piece of code. An entity could be an action (query/API), a JS function, a widget, or an `appsmith` object reference. That last type is not considered in our dependency logic today so we'll skip talking about it.

However, this approach was ridden with bugs because this meant that if we had name clashes inside comments, strings, local variables, etc., we would still continue to recognize those as a reference to the global entity. A smaller concern that arose from this was additional runs of queries/APIs. But it was also possible that we end up creating cyclical references along the way, even the the user had done no such thing.

With this set of changes, we will now have a more reliable way of figuring out these entity dependencies. Now, we are relying on the AST representation of any piece of code to find proper global variable references. This essentially means we'll be ignoring all the comment/string/local variable/function parameter references that were identified before. This also brings us closer to how the client calculates dependencies and hence brings consistency in these calculations.

Outside of this, the PR also introduces a few optimizations to the way we were processing these references that we had identified. Before this PR, we did not have a sense of the nature of each of these references, and in the process of calculating the dependency graph, we needed to refer to the DB over and over to check for this information. Now, we have introduces types that retain knowledge of whether a particular reference is a candidate for a widget, action or JS action reference. We also attach a little more information about the type of reference in terms of whether it is an async function, a direct function call, etc. Please check the `EntityDependencyNode` class that captures this information. 

Any identified node is propagated through the logic with this information so that we can rely on this one source of information for further processing.

This solution makes use of APIs exposed on the RTS server to calculate the on page load dependencies. Unfortunately, older slim container setups would not be able to accommodate this change without manual intervention from admins to set up RTS and then expose it to the server. As a result, this PR continues to use the older way of doing things for such set ups. It is highly recommended that users of Appsmith upgrade to the fat container set up to ensure a seamless experience.

TL;DR, here is the change set:
1. On page load calculations will now be more accurate. Good bye, a-certain-type-of cyclical dependency errors.
2. This change is only applicable for fat container setups. If you are not on this, please consider migrating.

Fixes #16338 
Fixes #9864 
Fixes #17190

## Type of change
- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?
- JUnit tests
- Manual testing

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/AstServiceCEImpl.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+@Slf4j
+@RequiredArgsConstructor
+public class AstServiceCEImpl implements AstServiceCE {
+
+    private final CommonConfig commonConfig;
+
+    private final InstanceConfig instanceConfig;
+
+    @Override
+    public Mono<List<String>> getPossibleReferencesFromDynamicBinding(String bindingValue, int evalVersion) {
```
[the remark this answers is not in the document]
Comment (nidhi-nair) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/AstServiceCEImpl.java:32]: This is the evaluated value actually. Any suggestions on how to improve the naming here?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/domains/ActionDependencyEdge.java [hunk 2]
```diff
@@ -12,14 +13,15 @@
...
+    EntityDependencyNode targetNode;
 
     @Override
     public int hashCode() {
         return new HashCodeBuilder()
-                .append(source)
-                .append(target)
+                .append(sourceNode.getReferenceString())
+                .append(targetNode.getReferenceString())
                 .toHashCode();
```
[the remark this answers is not in the document]
Comment (nidhi-nair) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/domains/ActionDependencyEdge.java:25]: Good question, the HashCodeBuilder is stateful and keeps the ordering of appended strings in mind, so this would not happen. 🙂

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/AstServiceCEImpl.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (subrata71) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/AstServiceCEImpl.java:32]: For a dynamic binding like `{{Api1.data}}` where the value of this is

`Api1.data = { "name": "x", "age": y}`

Do we call it like this?
```
bindingValue = "{ "name": "x", "age": y}";
evalVersion = 2
getPossibleReferencesFromDynamicBinding(bindingValue, evalVersion)
```
@nidhi-nair

[the remark this answers is not in the document]
Comment (nidhi-nair) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/AstServiceCEImpl.java:32]: Yes, stringified version though, so: 
```
bindingValue = "{\ "name\": \"x\", \"age\": y}";
```

File: app/server/appsmith-server/src/main/java/com/appsmith/server/configurations/CommonConfig.java [hunk 3]
```diff
@@ -62,8 +62,11 @@ public class CommonConfig {
     @Value("${disable.telemetry:true}")
     private boolean isTelemetryDisabled;
 
+    @Value("${appsmith.rts.url:http://localhost:8091}")
```
[the remark this answers is not in the document]
Comment (nidhi-nair) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/configurations/CommonConfig.java:65]: Yes this is intentional because we do not know what other things need to be handled in order to allow RTS to be supported at a different port at this point. I've just left an extension point for when we do get to using this feature.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/configurations/InstanceConfig.java [hunk 4]
```diff
@@ -76,18 +82,32 @@ private Mono<? extends Config> registerInstance() {
...
     }
 
+    private Mono<Void> performRtsHealthCheck() {
+        return WebClientUtils
+                .create(commonConfig.getRtsBaseDomain() + "/rts-api/v1/health-check")
+                .get()
+                .retrieve()
+                .toBodilessEntity()
+                .doOnNext(nextSignal -> this.isRtsAccessible = true)
+                .then();
```
[the remark this answers is not in the document]
Comment (nidhi-nair) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/configurations/InstanceConfig.java:92]: I don't think we should do periodic checks, simply for the reason that based on availability the behaviour that users see on the app will keep changing without their knowledge. It would give them suboptimal UX. 

For case 1, we'll be making use of Manish's changes to introduce these as error messages to be displayed on the UI so that users can contact their admins and correct the situation. 

For case 2, IMO there isn't much we can do about this. RTS is guaranteed to come up before the check because of our migration runs. But if it fails at that point, there's not much to be done.

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/models/EntityDependencyNode.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    String referenceString;
+    Boolean isAsync;
+    Boolean isFunctionCall;
+    ActionDTO actionDTO;
+
+    public boolean isValidDynamicBinding() {
+        boolean result = true;
+
+        if (this.referenceString == null) {
+            // This binding represents an on page load reference
```
[the remark this answers is not in the document]
Comment (nidhi-nair) [reply app/server/appsmith-interfaces/src/main/java/com/appsmith/external/models/EntityDependencyNode.java:24]: I meant actions that were manually set by the user to run on page load. I'll update the comment to make this clearer!

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/models/EntityDependencyNode.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        if (this.referenceString == null) {
+            // This node represents an entity that has been explicitly marked to run on page load
+            return true;
+        }
+
+        if (EntityReferenceType.ACTION.equals(this.entityReferenceType)) {
+            if (!this.referenceString.startsWith(this.validEntityName + ".data")) {
+                result = false;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (nidhi-nair) [reply app/server/appsmith-interfaces/src/main/java/com/appsmith/external/models/EntityDependencyNode.java:32]: This was a readability tradeoff. But I think I can add an `elif` for the JSAction block, nice catch! :)

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/AstServiceCEImpl.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        }
+
+        return WebClientUtils.create(commonConfig.getRtsBaseDomain() + "/rts-api/v1/ast/single-script-data")
+                .post()
+                .contentType(MediaType.APPLICATION_JSON)
+                .body(BodyInserters.fromValue(new GetIdentifiersRequest(bindingValue, evalVersion)))
+                .retrieve()
+                .bodyToMono(GetIdentifiersResponse.class)
+                .map(response -> response.data.references);
+        // TODO: add error handling scenario for when RTS is not accessible in fat container
```
[the remark this answers is not in the document]
Comment (nidhi-nair) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/AstServiceCEImpl.java:50]: Since this is not an `AppsmithException` this won't get caught in that block. We do need to add that path however. But we'll need client side changes to limit the number of times we show that message, hence I've put that change off to a subsequent PR. Also because such an occurrence will be rare.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/domains/ActionDependencyEdge.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (subrata71) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/domains/ActionDependencyEdge.java:25]: Will the `hashCodes` be different for the sample `referenceStrings` that I mentioned above?

[the remark this answers is not in the document]
Comment (nidhi-nair) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/domains/ActionDependencyEdge.java:25]: Yes they will be different for the two cases you have mentioned.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/AstServiceCEImpl.java [hunk 8]
```diff
@@ -66,6 +64,23 @@ static class GetIdentifiersResponse {
...
+     *     let c = "ignoredAction3.data"
+     *     // ignoredAction4.data
+     *     return aPostAction.data
+     *   } )(anotherPostAction.data)
+     * <p/>
+     * The values in the returned instance of GetIdentifiersResponseDetails will be:
+     * {
+     *   references: ["aPostAction.data", "anotherPostAction.data"],
+     *   functionalParams: ["ignoredAction1"],
+     *   variables: ["ignoredAction2"]
```
Comment (subrata71) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/AstServiceCEImpl.java:82]: Why are not `a,b,c` also returned as variables?

Comment (nidhi-nair) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/AstServiceCEImpl.java:82]: My bad, I only entered the ones that would be useful to us. I'll correct this. :)

**Comment by sondermanish** [comment]
LGTM

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/helpers/MustacheHelper.java [hunk 9]
```diff
@@ -322,21 +333,143 @@ public static String render(String template, Map<String, String> keyValueMap) {
...
+        }
+
+        if (subStrings.length == 2 && !"data".equals(subStrings[1])) {
+            // This could only qualify if it is a sync JS function call
+            // For sync JS actions, the entire reference could be a function call
+            EntityDependencyNode entityDependencyNode = new EntityDependencyNode(EntityReferenceType.JSACTION, key, reference, false, true, null);
+            dependencyNodes.add(entityDependencyNode);
+        }
+
+        if (subStrings.length >= 2) {
```
Comment (subrata71) [inline app/server/appsmith-interfaces/src/main/java/com/appsmith/external/helpers/MustacheHelper.java:427]: Couldn't it be an `elif`?

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/helpers/MustacheHelper.java [hunk 10]
```diff
@@ -322,21 +333,143 @@ public static String render(String template, Map<String, String> keyValueMap) {
...
+        String key = reference.trim();
+
+
+        String[] subStrings = nestedPathTokenSplitter.split(key);
+
+        if (subStrings.length < 1) {
+            return dependencyNodes;
+        }
+
+        if (subStrings.length == 2 && !"data".equals(subStrings[1])) {
```
Comment (subrata71) [inline app/server/appsmith-interfaces/src/main/java/com/appsmith/external/helpers/MustacheHelper.java:420]: Let's say I have a JS action like `JSObject1.data()`. In that case, won't we end up marking it as an `ACTION` instead of `JSACTION`? @nidhi-nair

Comment (nidhi-nair) [reply app/server/appsmith-interfaces/src/main/java/com/appsmith/external/helpers/MustacheHelper.java:420]: A function called `data`? I'm really not sure what the implications of that would be elsewhere in our universe. 🤔 Worth trying this out though, let me take a look.

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/helpers/MustacheHelper.java (hunk 9 above)
Comment (nidhi-nair) [reply app/server/appsmith-interfaces/src/main/java/com/appsmith/external/helpers/MustacheHelper.java:427]: Yes with my current code that makes sense. Let me check on the other thing and we'll sort this.

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/helpers/MustacheHelper.java (hunk 10 above)
Comment (nidhi-nair) [reply app/server/appsmith-interfaces/src/main/java/com/appsmith/external/helpers/MustacheHelper.java:420]: Okay seems to work okay from the on page load perspective at least. We'll also need to add a check to make sure that it _only_ gets identified as a JSAction in that scenario. I'll add that check and a few test cases as well. Thanks for catching this!

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/helpers/MustacheHelper.java (hunk 9 above)
Comment (nidhi-nair) [reply app/server/appsmith-interfaces/src/main/java/com/appsmith/external/helpers/MustacheHelper.java:427]: With the other change this will move around a little bit. Let me redo this.

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/helpers/MustacheHelper.java [hunk 11]
```diff
@@ -322,21 +333,149 @@ public static String render(String template, Map<String, String> keyValueMap) {
...
+     * Eg: JsObject2.asyncJsFunc.data which could be derived from {{ JsObject2.asyncFunc.data }}
+     * <p>
+     * Please note that we do not filter out any invalid references at this point,
+     * because we do not have context of the global namespace here.
+     *
+     * @param reference The string reference computed by AST logic
+     * @return A set of all possible action or JS function dependencies that could exist in the given reference
+     */
+    public static Set<EntityDependencyNode> getPossibleActions(String reference) {
+        Set<EntityDependencyNode> dependencyNodes = new HashSet<>();
```
Comment (subrata71) [inline app/server/appsmith-interfaces/src/main/java/com/appsmith/external/helpers/MustacheHelper.java:410]: Any reason, why it's a set of `EntityDependencyNode` instead of just `EntityDependencyNode`?

**Review by subrata71 — DISMISSED** [review]
The PR looks good to me.

**Review by sondermanish — DISMISSED** [review]

**Comment by nidhi-nair** [comment]
Added Test plan approved label since @Parthvi12 and @Aishwarya-U-R have given sign off on the PR.

**Review by sondermanish — APPROVED** [review]
